src/MinimalDelegation.sol

```diff
@@ -144,55 +144,49 @@ contract MinimalDelegation is
         returns (uint256 validationData)
     {
         _payEntryPoint(missingAccountFunds);
-        (bytes32 keyHash, bytes memory signature) = abi.decode(userOp.signature, (bytes32, bytes));
+        (bytes32 keyHash, bytes memory signature, bytes memory witness) = abi.decode(userOp.signature, (bytes32, bytes, bytes));
 
+        Key memory key = getKey(keyHash);
         Settings settings = getKeySettings(keyHash);
         (bool isExpired, uint40 expiry) = settings.isExpired();
         if (isExpired) revert IKeyManagement.KeyExpired(expiry);
 
-        IHook hook = settings.hook();
-        validationData = hook.hasPermission(HooksLib.VALIDATE_USER_OP_FLAG)
-            ? hook.validateUserOp(keyHash, userOp, userOpHash)
-            : _handleValidateUserOp(keyHash, signature, userOp, userOpHash, expiry);
-    }
-
-    /// TODO: This is left as an internal function to handle wrapping the returned validation data accoring to ERC-4337 spec.
-    function _handleValidateUserOp(
-        bytes32 keyHash,
-        bytes memory signature,
-        PackedUserOperation memory,
-        bytes32 userOpHash,
-        uint40 expiry
-    ) private view returns (uint256 validationData) {
-        Key memory key = getKey(keyHash);
         /// The userOpHash does not need to be safe hashed with _hashTypedData, as the EntryPoint will always call the sender contract of the UserOperation for validation.
         /// It is possible that the signature is a wrapped signature, so any supported key can be used to validate the signature.
         /// This is because the signature field is not defined by the protocol, but by the account implementation. See https://eips.ethereum.org/EIPS/eip-4337#definitions
 
         /// validationData is (uint256(validAfter) << (160 + 48)) | (uint256(validUntil) << 160) | (success ? 0 : 1)
         /// `validAfter` is always 0.
-        if (key.verify(userOpHash, signature)) return uint256(expiry) << 160 | SIG_VALIDATION_SUCCEEDED;
-        else return uint256(expiry) << 160 | SIG_VALIDATION_FAILED;
+        if(key.verify(userOpHash, signature)) {
+            validationData = uint256(expiry) << 160 | SIG_VALIDATION_SUCCEEDED;
+        } else {
+            validationData = uint256(expiry) << 160 | SIG_VALIDATION_FAILED;
+        }
+
+        IHook hook = settings.hook();
+        if(hook.hasPermission(HooksLib.VALIDATE_USER_OP_FLAG)) {
+            // Will revert if validation should fail
+            hook.checkValidateUserOp(keyHash, userOp, userOpHash, witness);
```

Comment:
should hook be able to override the validationData ? 

ie maybe it puts a tighter deadline on validity?